File: c/changepw.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "sillysite.h"
#include "readpass.h"

int main(int argc, char *argv[])
{
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <baseurl> <username>\n", argv[0]);
        return 1;
    }

    const char *baseurl  = argv[1];
    const char *username = argv[2];

    char oldpw[1024], newpw[1024], confirmpw[1024];

    if (read_password("Current password: ",     oldpw,     sizeof(oldpw))     < 0 ||
        read_password("New password: ",         newpw,     sizeof(newpw))     < 0 ||
        read_password("Confirm new password: ", confirmpw, sizeof(confirmpw)) < 0) {
        fprintf(stderr, "Failed to read password\n");
        return 1;
    }

    if (strcmp(newpw, confirmpw) != 0) {
        memset(oldpw, 0, sizeof(oldpw));
        memset(newpw, 0, sizeof(newpw));
        memset(confirmpw, 0, sizeof(confirmpw));
        fprintf(stderr, "Change password failed: passwords do not match\n");
        return 1;
    }
    memset(confirmpw, 0, sizeof(confirmpw));

    int rc = silly_changepw(baseurl, username, oldpw, newpw);
    memset(oldpw, 0, sizeof(oldpw));
    memset(newpw, 0, sizeof(newpw));

    if (rc != 0) {
        fprintf(stderr, "Change password failed: %s\n", strerror(errno));
        return 1;
    }

    printf("Password changed successfully\n");
    return 0;
}

File: php/changepw.php
#!/usr/bin/env php
<?php
/**
 * Usage: ./changepw.php <baseurl> <username>
 *
 * Prompts for the current password, a new password, and a confirmation.
 * Verifies the two new-password entries match before proceeding. Mirrors
 * ../changepw.py, ../c/changepw.c, ../js/changepw.js, ../java/ChangePw.java,
 * and ../csharp/ChangePw.cs.
 */
require __DIR__ . "/Sillysite.php";
require __DIR__ . "/Readpass.php";

function main(): int
{
    global $argv;
    if (count($argv) !== 3) {
        fwrite(STDERR, "Usage: " . basename($argv[0]) . " <baseurl> <username>\n");
        return 1;
    }
    [, $baseUrl, $username] = $argv;

    $oldPassword = readPassword("Current password: ");
    $newPassword = readPassword("New password: ");
    $confirmPassword = readPassword("Confirm new password: ");

    if ($newPassword !== $confirmPassword) {
        fwrite(STDERR, "Change password failed: passwords do not match\n");
        return 1;
    }

    try {
        Sillysite::changepw($baseUrl, $username, $oldPassword, $newPassword);
        echo "Password changed successfully\n";
        return 0;
    } catch (SillysiteException $e) {
        fwrite(STDERR, "Change password failed: " . $e->getMessage() . "\n");
        return 1;
    }
}

exit(main());
